### ISSUE TITLE
Generalize GHA `select` statements (to avoid hard-coding versions)

### ISSUE THREAD

**Issue by jakirkham** [rapidsai/build-planning#25]
As there is only need of Dask-CUDA build, we filter out one architecture, Python version, and CUDA version in GHA. So as to only build the package once

https://github.com/rapidsai/dask-cuda/blob/1eecb1b2ac79ae9aaff9c26d0a3c93dd57f859f3/.github/workflows/build.yaml#L69-L70

However this currently hard-codes the versions of each in the selection logic, which means it can go stale as new versions are added and old ones dropped. Potentially resulting in the build being lost altogether (maybe even silently)

To avoid pinning to a specific version, @ajschmidt8 made several suggestions in this thread: https://github.com/rapidsai/dask-cuda/pull/1294#discussion_r1417857509

Filing this to track for follow-up

**Comment by jakirkham** [comment]
Here is another hard-coding example in cuDF

https://github.com/rapidsai/cudf/blob/3adfddcfa2cdac4acb16a50916442763a1d8a78b/.github/workflows/build.yaml#L94-L95

**Comment by jakirkham** [comment]
Also some relevant discussion in this thread: https://github.com/rapidsai/cudf/pull/15174#discussion_r1506731985

**Comment by jameslamb** [comment]
@bdice I tried the "Pure Python, no CUDA dependency" example in https://github.com/rapidsai/rapids-dask-dependency/pull/28

Worked perfectly, thank you 😊

**Comment by jakirkham** [comment]
Also needed in cuxfilter

https://github.com/rapidsai/cuxfilter/blob/cfab998ffc4f348f649cd839bea5697fd8aeef02/.github/workflows/pr.yaml#L67

**Comment by bdice** [comment]
I did cudf as well. https://github.com/rapidsai/cudf/pull/15191

There are some hardcoded references remaining but they're not really used in the same way as the cases we've addressed above.

Docker needs something like `if latest CUDA` but there's no matrix filter so I'm not sure if there's a good solution.
https://github.com/rapidsai/docker/blob/32a60e107c3ff777c7f2ddc2ac4d11a3669c3f83/.github/workflows/build-image.yml#L145

cuGraph has some PyG tests that only run on CUDA 11. I don't think these are worth modifying since it's pinning to an older version. There is some open work needed to upgrade to PyTorch 2, which I think may open up the ability to use CUDA 12 here.
https://github.com/rapidsai/cugraph/blob/ac65b17ee3e9b85368f266da1a6a3b8e5717e292/.github/workflows/pr.yaml#L165
https://github.com/rapidsai/cugraph/blob/ac65b17ee3e9b85368f266da1a6a3b8e5717e292/.github/workflows/test.yaml#L79
